### PR TITLE
Remove Hive/Hadoop/Java dependency from unit tests

### LINKED ISSUE #8783: Move Hadoop distro out of CI image
**Description**

Hadoop distro is unnecessarily part of the CI image. It should be moved out and used separately

**Use case / motivation**

Download and install Hadoop image only when needed not as part of the CI image (as separate integration). This will make the image smaller and the whole process smoother.

### PR DESCRIPTION
**Pull request by jhtimmins**
CI takes way too long to build and uses too much memory. This removes Hive and Hadoop as dependencies from CI, and replaces direct access with mocks in unit tests.

Fixes #8783 (though differently to proposed, via mocking instead of separate docker image)
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by ashb** [comment]
Looks like the tests are (mostly) passing now :tada:

File: Dockerfile.ci
Comment (ashb) [inline Dockerfile.ci]: What still needs the JRE btw?

File: airflow/providers/apache/hive/operators/hive.py
Comment (ashb) [inline airflow/providers/apache/hive/operators/hive.py]: What's this for? Perhaps a comment explaining why we do this might be helpful.

File: tests/providers/apache/hive/operators/test_hive_to_mysql.py
Comment (ashb) [inline tests/providers/apache/hive/operators/test_hive_to_mysql.py]: What's it doing? Where is it doing it?

File: Dockerfile.ci
Comment (jhtimmins) [reply Dockerfile.ci]: Removed

**Comment by jhtimmins** [comment]
@mik-laj This is now failing tests that seem related to this change you made to `test_scheduler_job.py` (https://github.com/apache/airflow/commit/93b8f3e48df0a1aaa7be4f83904f4208496ec7db#). It's only failing on CI for some of the environments, https://github.com/apache/airflow/runs/732110359?check_suite_focus=true#step:6:347.

Can you provide any insight into the changes you made and they proper approach if those tests are failing?

**Comment by mik-laj** [comment]
@jhtimmins My change has caused us to cross a certain threshold that causes the exact cause of this instability. I will want to track her cause soon.  I restarted Github Action build. That should help now.
In a moment I will add this test to quarantine to give myself more time to look at it.

**Comment by jhtimmins** [comment]
@mik-laj Awesome, that's good to know. Thanks for the info!

File: tests/providers/apache/hive/hooks/test_hive.py
Comment (kaxil) [inline tests/providers/apache/hive/hooks/test_hive.py:669]: Why do we assert this call over here?

Comment (jhtimmins) [reply tests/providers/apache/hive/hooks/test_hive.py:669]: Generally speaking, I tried to mock assert all calls to external systems. These test that following line is run. https://github.com/apache/airflow/blob/master/airflow/providers/apache/hive/hooks/hive.py#L843

**Review by ashb — APPROVED** [review]
This LGTM.

@mik-laj Should we merge even with the (unrelated to this PR) query count test failing?

**Comment by potiuk** [comment]
> This LGTM.
> 
> @mik-laj Should we merge even with the (unrelated to this PR) query count test failing?

@mik-laj  is on holidays but  I know this test is flaky and he planned to move it to quarantine or fix . so yea. let's

**Comment by kaxil** [comment]
I think we can merge this one if Master fails once this PR is merged, we can add that test to quarantine to allow Kamil some time to look at it when he is free. (He is on holiday today)

**Comment by potiuk** [comment]
Fantastic @jhtimmins ! This will be really, really helpful and get our CI  image  much smaller and build process much faster! :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

**Comment by kaxil** [comment]
Good work @jhtimmins 👏 👏

**Comment by ashb** [comment]
Shrunk down the size of the image by 1GB!

```
docker.pkg.github.com/apache/airflow/master-python3.6-ci   latest                             8996f52da18d        35 minutes ago      3.43GB
apache/airflow                                             master-python3.6-ci                30378708ee4c        4 days ago          4.44GB
```

(Don't trust sizes docker hub reports -- it "lies" and gives the compressed size.)

**Comment by dimberman** [comment]
Thank you @jhtimmins! I owe you a soda when Los Angeles opens up again (so... maybe by 2022)

**Comment by potiuk** [comment]
Heyll yeah!

**Comment by potiuk** [comment]
Hey @jhtimmins -> I cherry-picked your changes to https://github.com/apache/airflow/commit/d4a00b7c69d5d1c4514bc70b2a0022f36b85a06b 

I'd love if you can take a look and see if I figured the cherry-picking well :)

**Pull request merged**
